### ISSUE TITLE
bug in initGpmonPktFuncs[] (execUtils.c)? Dead Code?

### ISSUE THREAD

**Issue by CraigHarris** [greenplum-db/gpdb#256]
initGpmonPktFuncs is a jump table, whose size and order of entries is supposed to match the plan-node types in nodes/nodes.h.  But it is missing the entry for TableFunctionScan (between FunctionScan and ValueScan), along with the entries for PlanNode types after T_Repeat.

sendInitGpmonPkts uses the expression [nodeTag(node) - T_Plan_Start] as an index into the jump table, so about half the plan types (those after T_FunctionScan) will jump to the wrong init function. Even worse, plans at T_Repeat and after will jump off the end of the table.

There is (supposedly) a compile-time check that the number of entries in the table matches T_Plan_End - T_Plan_Start.  But that doesn't guarantee order correctness, nor does it appear to be working as intended.

But it looks like this whole ball of wax is dead code. sendInitGpmonPkts calls itself, but nothing seems to call it. 

So... either the dead code should be removed (including the initGpmonPktForXXX functions in each executor/nodeType.c file), or an entry for TableFunctionScan should be added to the initGpmonPktFuncs jump table

**Comment by hlinnaka** [comment]
On 09/01/16 18:01, Craig Harris wrote:
> initGpmonPktFuncs is a jump table, whose size and order of entries is
> supposed to match the plan-node types in nodes/nodes.h.  But it is
> missing the entry for TableFunctionScan (between FunctionScan and
> ValueScan), along with the entries for PlanNode types after
> T_Repeat.

Good catch.

> sendInitGpmonPkts uses the expression [nodeTag(node) - T_Plan_Start]
> as an index into the jump table, so about half the plan types (those
> after T_FunctionScan) will jump to the wrong init function. Even
> worse, plans at T_Repeat and after will jump off the end of the
> table.
>
> There is (supposedly) a compile-time check that the number of entries
> in the table matches T_Plan_End - T_Plan_Start.  But that doesn't
> guarantee order correctness, nor does it appear to be working as
> intended.
>
> But it looks like this whole ball of wax is dead code.
> sendInitGpmonPkts calls itself, but nothing seems to call it.
>
> So... either the dead code should be removed (including the
> initGpmonPktForXXX functions in each executor/nodeType.c file), or an
> entry for TableFunctionScan should be added to the initGpmonPktFuncs
> jump table

I'm inclined to remove the whole thing. The comment in sendInitGpmonPkts 
says that it's only called by "the Append executor node, since Append 
does not call ExecInitNode() for each of its child nodes during 
initialization time". But looking at ExecInitAppend(), it does seem to 
call ExecInitNode on each of its children. So I think the comment is 
out-of-date, and the function is no longer needed.

Want to write a patch to remove it all?

To be honest though, I don't know what gpmon is. Anyone? And can someone 
verify whether gpmon is currently working correctly with Append plans?

- Heikki

**Comment by hlinnaka** [comment]
On 09/01/16 22:08, Heikki Linnakangas wrote:
> The comment in sendInitGpmonPkts says that it's only called by "the
> Append executor node, since Append does not call ExecInitNode() for
> each of its child nodes during initialization time". But looking at
> ExecInitAppend(), it does seem to call ExecInitNode on each of its
> children. So I think the comment is out-of-date, and the function is
> no longer needed.

I dug into old git history of this. sendInitGpmonPkts() was added in a 
comment that changed ExecInitAppend() to postpone the call to 
ExecInitNode() to child nodes, under some circumstances. 
sendInitGpmonPkts() was called instead of ExecInitNode(). But that 
commit was reverted a few weeks later, because it apparently caused 
problems with EXPLAIN ANALYZE. However, sendInitGpmonPkts() was left 
behind and not removed by the reverting commit.

So that code is dead as a doornail.

- Heikki

**Comment by ivannovick** [comment]
For history, let me explain a little background.

We have gpperfmon agents that collect information from GPDB and store it in tables.  There are several primary tables:

system_history: CPU, disk, network, memory utilitiation, etc.  No interaction with GPDB itself, just the OS

database_history: records number of querys active, queued, running

query_history: records each query run and the query text.  This is a critical functionality and is generated from the master segment.  The design of it has always been a bit troublesome and needs to be re-designed at some point because it can sometimes from time to time miss a query and users want the query history to be 100% accurate.  Also it uses flat files on the master as a communication mechanism which creates too many flat files in some cases, although they are cleaned, its still a clunky mechnanism.  That being said, we need a new solution before we can remove querys_history generatiion.

iterators_history.  This records information periodically from running queries from each plan node on each segment.  This allows you to track the progress of running queries at the plan node level within each segment.  This is the majority of the code in GPDB related to the gpmon packets.  This feature has been very troublesome because of numerous reasons including pumping out too granular data that can overload the master and waste memory and disk space.  Also difficulty keeping all the executor nodes code sending the right metrics has been a burden on developers, and visualizing all this is difficult on the UI.

My suggestion is to obsolete iterator_history data publishing in Greenplum 5 but we really need to keep (or redesign) the mechanism for outputing the query_history data to allow UI's to see currently running queries with some metrics and to provide an audit log of the history of all queries.

Probably i left out some details but this is the general background

**Comment by ivannovick** [comment]
also, last time i checked, i forget if Pivotal open sourced the gpmmon and gpsmon agents, but that needs to be open sourced and added to the build if its not already IMO.

**Comment by CraigHarris** [comment]
Seems like cleaning this up isn't urgent, and nobody's at risk of jumping off the end of the function-ptr table.  I'll submit a pull request late next week or the following to remove the dead code.

**Comment by hlinnaka** [comment]
I created a pull request to remove the dead code, so that this doesn't get forgotten about.

**Comment by hlinnaka** [comment]
Ok, pushed the removal of that, closing this issue. Thanks for the report!
